target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,750 @@
+#define COMPAT_RTL839X "realtek,rtl8390-clock"
+
+#define CLK_INFO(_idx, _name, _pname, _dname, _dtrange)	\
+	{						\
+		.idx = _idx,				\
+		.name = _name,				\
+		.parent_name = _pname,			\
+		.display_name = _dname,			\
+		.dtrange = _dtrange			\
+	}
+
+struct rtcl_clk_info {
+	unsigned int idx;
+	const char *name;
+	const char *parent_name;
+	const char *display_name;
+	const char *dtrange;
+};
+
+struct rtcl_clk {
+	unsigned int idx;
+	unsigned long min;
+	unsigned long max;
+	unsigned long startup;
+	struct clk_hw hw;
+};
+
+static const struct rtcl_clk_info rtcl_clk_info[CLK_COUNT] = {
+	CLK_INFO(CLK_CPU, "cpu_clk", "ref_clk", "CPU", "cpu-range"),
+	CLK_INFO(CLK_MEM, "mem_clk", "ref_clk", "MEM", "mem-range"),
+	CLK_INFO(CLK_LXB, "lxb_clk", "ref_clk", "LXB", "lxb-range")
+};
+
+struct rtcl_dram {
+	int type;
+	int buswidth;
+};
+
+struct rtcl_sram {
+	unsigned long vbase;
+	int *pmark;
+};
+
+struct rtcl_ccu {
+	int soc;
+	struct rtcl_sram sram;
+	struct rtcl_dram dram;
+	bool dynamic;
+	spinlock_t lock;
+	struct platform_device *pdev;
+	struct device_node *np;
+	struct rtcl_clk clks[CLK_COUNT];
+};
+
+struct rtcl_ccu *rtcl_ccu;
+
+#define rtcl_hw_to_clk(_hw) container_of(_hw, struct rtcl_clk, hw)
+
+/*
+ * SRAM relocatable assembler functions
+ */
+
+extern void	rtcl_838x_dram_start(void);
+extern int	rtcl_838x_dram_size;
+
+extern void	(*rtcl_838x_dram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+static void	(*rtcl_838x_sram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+
+extern void	(*rtcl_839x_dram_set_rate_mem)(int clk_idx, int ctrl0, int ctrl1);
+static void	(*rtcl_839x_sram_set_rate_mem)(int clk_idx, int ctrl0, int ctrl1);
+
+extern void	rtcl_839x_dram_start(void);
+extern int	rtcl_839x_dram_size;
+
+extern void	(*rtcl_839x_dram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+static void	(*rtcl_839x_sram_set_rate)(int clk_idx, int ctrl0, int ctrl1);
+
+/*
+ * clock setter/getter functions
+ */
+
+static unsigned long rtcl_recalc_rate(struct clk_hw *hw, unsigned long parent_rate)
+{
+	struct rtcl_clk *clk = rtcl_hw_to_clk(hw);
+
+	unsigned int ctrl0 = read_sw(rtcl_regs[rtcl_ccu->soc][REG_CTRL0][clk->idx]);
+	unsigned int ctrl1 = read_sw(rtcl_regs[rtcl_ccu->soc][REG_CTRL1][clk->idx]);
+
+	unsigned int div1, div2 = rtcl_xdiv[clk->idx];
+	unsigned int cmu_sel_prediv = 1 << RTL_PLL_CTRL0_CMU_SEL_PREDIV(ctrl0);
+	unsigned int cmu_sel_div4 = RTL_PLL_CTRL0_CMU_SEL_DIV4(ctrl0) ? 4 : 1;
+	unsigned int cmu_ncode_in = RTL_PLL_CTRL0_CMU_NCODE_IN(ctrl0) + 4;
+	unsigned int cmu_divn2 = RTL_PLL_CTRL0_CMU_DIVN2(ctrl0) + 4;
+	unsigned int cmu_divn2_selb, cmu_divn3_sel;
+
+	switch (rtcl_ccu->soc) {
+	case SOC_RTL838X:
+		cmu_divn2_selb = RTL838X_PLL_CTRL1_CMU_DIVN2_SELB(ctrl1);
+		cmu_divn3_sel = rtcl_divn3[RTL838X_PLL_CTRL1_CMU_DIVN3_SEL(ctrl1)];
+		break;
+	case SOC_RTL839X:
+		cmu_divn2_selb = RTL839X_PLL_CTRL1_CMU_DIVN2_SELB(ctrl1);
+		cmu_divn3_sel = rtcl_divn3[RTL839X_PLL_CTRL1_CMU_DIVN3_SEL(ctrl1)];
+		break;
```

Comment:
```suggestion
		break;
	default:
		dev_err(hw->dev, "Unable to set rate '%d' for unknown SoC ID '%d'\n", rate, rtcl_ccu->soc);
		return -EINVAL;
```

Reply:
This will not happen as rtcl_ccu_create() will fail during initialization.

Reply:
But that's making assumptions; stuff may get refactored, and this function may be called with different understanding context. Being defensive here, adds 1/2 lines of code, pleases some linters (all cases caught) and potentially catches a 1 in a million bug.